Email Editor: Integrate the convert class from the HTML to Text library

allilevine opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

* Closes https://linear.app/a8c/issue/WOOPLUG-4979/integrate-the-html-to-text-library-into-the-email-editor
* Removes the `soundasleep/html2text` dependency.
* Forks the `Html2Text::convert` method from `soundasleep/html2text` with WordPress code formatting.
* Adds unit tests from `soundasleep/html2text` and extends the existing Renderer integration tests.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

- [ ] Check that tests pass.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

Integrated the convert class from the external HTML to Text library and removed the dependency from the Email Editor package.

</details>

allilevine committed:
# This is a combination of 5 commits.
# This is the 1st commit message:

Email Editor: Integrate the convert class from the HTML to Text library and remove the dependency.

# This is the commit message #2:

Fix linting errors.

# This is the commit message #3:

Fix more linting errors.

# This is the commit message #4:

Fix package linting errors.

# This is the commit message #5:

Fix additional package linting errors.

allilevine committed:
Email Editor: Integrate HTML to Text library and remove external dependency

- Internalized soundasleep/html2text functionality
- Added Html2Text and Html2Text_Exception classes
- Updated composer.json to remove dependency
- Fixed all PHPStan and WordPress coding standards violations
- Ensured PHP 7.4+ compatibility and standalone package support

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text.php:
```diff
@@ -0,0 +1,622 @@
+		// Check all options are valid.
+		foreach ( $options as $key => $value ) {
+			if ( ! in_array( $key, array_keys( static::default_options() ), true ) ) {
+				// phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
```
I'm assuming that this package needs to work standalone without WordPress core functions available, so using `htmlspecialchars()` instead of `esc_html()`. Let me know if that's an incorrect assumption.

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text.php:
```diff
@@ -0,0 +1,622 @@
+		}
+
+		if ( ! empty( $options['ignore_errors'] ) ) {
+			// phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
```
Ignoring snake case rule because these are PHP's built-in property names from the DOM classes.

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text.php:
```diff
@@ -0,0 +1,622 @@
+	 * @return bool True if this appears to be an Office document.
+	 */
+	public static function is_office_document( string $html ): bool {
+		return strpos( $html, 'urn:schemas-microsoft-com:office' ) !== false;
```
Using `strpos` instead of `str_contains` here (and `str_starts_with` elsewhere in the file) because of the PHP 7.4 compatability requirement.

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text.php:
```diff
@@ -0,0 +1,622 @@
+<?php
```
Original file for comparison: https://github.com/soundasleep/html2text/blob/master/src/Html2Text.php

allilevine committed:
Fix potential security issues with exceptions.

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text-exception.php:
```diff
@@ -0,0 +1,33 @@
+	 *
+	 * @var string
+	 */
+	public $more_info;
```
Fixed in https://github.com/woocommerce/woocommerce/pull/59859/commits/770d04df8d29e49b54676955d2cdc827620e816a

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text.php:
```diff
@@ -0,0 +1,622 @@
+		if ( ! $load_result ) {
+			// phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+			throw new Html2Text_Exception( 'Could not load HTML - badly formed?', htmlspecialchars( $html, ENT_QUOTES, 'UTF-8' ) );
+		}
```
Fixed in https://github.com/woocommerce/woocommerce/pull/59859/commits/770d04df8d29e49b54676955d2cdc827620e816a

allilevine commented on packages/php/email-editor/src/Engine/Renderer/class-html2text.php:
```diff
@@ -0,0 +1,622 @@
+		foreach ( $options as $key => $value ) {
+			if ( ! in_array( $key, array_keys( static::default_options() ), true ) ) {
+				// phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+				throw new \InvalidArgumentException( 'Unknown html2text option \'' . htmlspecialchars( $key, ENT_QUOTES, 'UTF-8' ) . '\'. Valid options are ' . htmlspecialchars( implode( ',', array_keys( static::default_options() ) ), ENT_QUOTES, 'UTF-8' ) );
```
Fixed in https://github.com/woocommerce/woocommerce/pull/59859/commits/770d04df8d29e49b54676955d2cdc827620e816a

allilevine committed:
Ignore error_logs since they're for security purposes.

allilevine committed:
Remove unused .

jeherve commented on packages/php/email-editor/src/Engine/Renderer/class-renderer.php:
```diff
@@ -12,7 +12,7 @@
 use Automattic\WooCommerce\EmailEditor\Engine\Theme_Controller;
 use Automattic\WooCommerce\EmailEditor\Engine\PersonalizationTags\Personalization_Tags_Registry;
-use Soundasleep\Html2Text;
+use Automattic\WooCommerce\EmailEditor\Engine\Renderer\Html2Text;
```
Since it's in the same namespace (`Automattic\WooCommerce\EmailEditor\Engine\Renderer`), I don't think we need the import here.


```suggestion
```

allilevine replied:
Thanks, you're right! Removed in https://github.com/woocommerce/woocommerce/pull/59859/commits/dda69faaff724605a74575f034516d86f8c1a1d5

jeherve reviewed (commented):
Would it be worth bringing some of the tests from `soundasleep/html2text`'s `tests/Html2TextTest.php` alongside with the class, to ensure there were no regressions?

Maybe we could also extend the tests in `packages/php/email-editor/tests/integration/Engine/Renderer/Renderer_Test.php` to check for HTML conversion? We do have a few existing tests that ensure things are working (like `testItRendersPostWithinAssociatedTemplate`), but maybe there could be an additional one with a data provider that would check for a few usual tags, like divs, links, and maybe special entities?

pavel-mailpoet reviewed (dismissed):
Great job @allilevine thank you for preparing the pull request. 

Apart from what Jeremy wrote, I think it would be good to include a changelog in this pull request. I saw your comment, but I still think we should add it. Otherwise, when we publish a new version of the library without a changelog, people wouldn’t know what the difference is between the versions.

jeherve commented:
> Do we need to include any original copyright lines and a copy of the MIT license to include this?

That's an excellent point, and a good question. 

Internal reference to figure that out: p2y3YZ-9Ry-p2

allilevine committed:
Email editor: Add unit and integration tests for html2text, and remove unnecessary import.

allilevine committed:
Fix PHP and linter issues.

allilevine committed:
Attempt to keep the tests happy while fixing the PHPStan error.

allilevine committed:
Change from double quotes to single quotes.

Commit:
Add changefile(s) from automation for the following project(s): packages/php/email-editor

jeherve committed:
Add security information to package

jeherve committed:
Add copyright and license headers to new files

jeherve committed:
Add package license

- GPLv2 or later
- Also incorporates work with different copyright and license.

Internal reference:
p2y3YZ-9Ry-p2#comment-22676

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into wooplug-4979-integrate-the-html-to-text-library-into-the-email-editor

jeherve committed:
Add new licensing / security reporting policy

kalessil commented on packages/php/email-editor/license.txt:
```diff
@@ -0,0 +1,365 @@
+This program is free software; you can redistribute it and/or modify
```
We need to add the new license and security files to the package mirror process: see https://github.com/woocommerce/woocommerce/blob/trunk/packages/php/email-editor/tasks/build-package.sh#L35

jeherve replied:
👍 Done in fc30ed5e074af45593c2c9b457c7d5d24b55fe15

kalessil commented on packages/php/email-editor/changelog/59859-wooplug-4979-integrate-the-html-to-text-library-into-the-email-editor-2:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+Add licensing and security policy.
```
No need for extra changelog entries; just update the first one accordingly.

jeherve replied:
Since the 2 items are not strictly related, and could have been handled in 2 different PRs (one adding the security and basic license file, the other editing the license file to add the new extra copyright), it seems like having 2 changelog entries may be cleaner, and maybe interesting for consumers of the package?

kalessil replied:
I see the point. Well, I don't have any good insight into the package consumers profile - feel free to ignore this comment =)

jeherve committed:
Add 2 new files to mirror repo as well

jeherve reviewed (approved):
This looks good to me, and the tests look good. I'll approve, but I'll let someone more familiar with the package chime in too :)

pavel-mailpoet reviewed (approved):
Thank you again for your contribution, tests are passing, so let's merge it @allilevine
